streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java

```diff
@@ -253,17 +272,18 @@ private void rebuildMetadata(final Map<HostInfo, Set<TopicPartition>> currentSta
                                                          final SourceTopicsInfo sourceTopicsInfo) {
 
         final Integer partition = partitioner.partition(sourceTopicsInfo.topicWithMostPartitions, key, null, sourceTopicsInfo.maxPartitions);
-        final Set<TopicPartition> matchingPartitions = new HashSet<>();
-        for (final String sourceTopic : sourceTopicsInfo.sourceTopics) {
-            matchingPartitions.add(new TopicPartition(sourceTopic, partition));
+        final Set<TaskId> matchingPartitions = new HashSet<>();
+        final Map<Integer, InternalTopologyBuilder.TopicsInfo> topicGroups = builder.topicGroups();
+        for (final Integer topicGroupId : topicGroups.keySet()) {
+            matchingPartitions.add(new TaskId(topicGroupId, partition));
         }
 
         for (final StreamsMetadata streamsMetadata : allMetadata) {
             final Set<String> stateStoreNames = streamsMetadata.stateStoreNames();
-            final Set<TopicPartition> topicPartitions = new HashSet<>(streamsMetadata.topicPartitions());
-            topicPartitions.retainAll(matchingPartitions);
+            final Set<TaskId> taskIds = new HashSet<>(streamsMetadata.taskIds());
```

Comment:
I think we can just keep the logic to maintain the topic-partition set and not exposing tasks instead, the key is that in `rebuildMetadata`, we can still reconstruct the topic-partition set. What I have in mind is the following:

1) we remember the `final Map<TaskId, Set<TopicPartition>> partitionsForTask = partitionGrouper.partitionGroups(sourceTopicsByGroup, fullMetadata);` constructed from StreamPartitionAssignor at the `StreamsMetadataState`. This can be pass via the `assignor -> taskManager -> streamMetadataState`, i.e. at `setPartitionsByHostState` we pass in both `tasksByHosts` and `partitionsForTask`.

2) And then at `rebuildMetadata` we can get the `Set<TopicPartition>` from `taskId` directly.

Reply:
What's the thought behind this, we don't want to change StreamMetadata i.e. keep partitionsByHost in the StreamsMetadata instead of tasksByHost?

Reply:
Also, final Map<TaskId, Set<TopicPartition>> partitionsForTask = partitionGrouper.partitionGroups(sourceTopicsByGroup, fullMetadata);

This information is present inside assign() function which is called only on cosumer coordinator. All other machines get assignment via onAssignment(). So,  you want to populate sourceTopicsByGroup and fullMetadata in onAssignment() as below?

final Map<Integer, InternalTopologyBuilder.TopicsInfo> topicGroups = taskManager.builder().topicGroups();

        final Set<String> allSourceTopics = new HashSet<>();
        final Map<Integer, Set<String>> sourceTopicsByGroup = new HashMap<>();
        for (final Map.Entry<Integer, InternalTopologyBuilder.TopicsInfo> entry : topicGroups.entrySet()) {
            allSourceTopics.addAll(entry.getValue().sourceTopics);
            sourceTopicsByGroup.put(entry.getKey(), entry.getValue().sourceTopics);
        }

        Cluster fullMetadata = Cluster.empty().withPartitions(topicToPartitionInfo))
        
        final Map<TaskId, Set<TopicPartition>> partitionsForTask = partitionGrouper.partitionGroups(sourceTopicsByGroup, fullMetadata);

        taskManager.setClusterMetadata(fullMetadata);

Reply:
@brary My fault for not being explicit enough :) Let me try again.

Your changes on the public API `StreamsMetadata` seems only used once, which is here to determine the owner of the partitions -- since we are removing the map from host to partitions, we need to expose the API that maps from host to tasks and construct the partitions from tasks.

What's in my mind is that, when we `rebuildMetadata`, we can still construct the partitions set as the constructor parameters. And the way to do that is:

1. at `onAssignment`, use the `builder` to first get `builder.topicGroups()`, and from topicGroups to get `sourceTopicsByGroup` (see the logic in the `assign` function under comments `// ---------------- Step One ---------------- //`). And then from `sourceTopicsByGroup`, for each `taskId` we can get its partitions list by appending its taskId's partition id to each of the topics.

**NOTE** that this actually bypassed the whole `partitionGrouper.partitionGroups` call by assuming that we always use the default grouper (i.e. it was not really "remember" the map but just re-execute the way it was generated). This is a bold assumption, but we are already making such assumptions anyways even today.

We should consider removing the partition grouper API completely, but that would be a different topic beyond this PR.

2. with that we can even slightly change our proposed protocol to the following to encode the task -> partitions mapping:

```
   ActiveTasks               => List<TaskId>
   StandbyTasks            => List<TaskId>    // here is the change
   PartitionsByHost      => Map<HostInfo, Set<TaskId>>
```

For standby tasks we can remove its associated source partitions as well since we can get them from 1) above as well.

WDYT?

Reply:
For now I am thinking of just executing the first step, I had thought of what you are proposing in step 2 and changing topic partitions to List<TaskId> for Standby tasks, but we can take this change later as an improvement since it won't have much effect in terms of size, as in large cluster size occupied by standby topic partitions of one machine won't be much.

Reply:
Sounds good. Thanks!